File: scripts/build_screen.js
#!/usr/bin/env node

import { parseArgs as parseNodeArgs } from "node:util";
import { resolve } from "node:path";
import { readCsv, writeCsvAtomic } from "./lib/csv.js";
import { daysBetween, parseDate } from "./lib/dates.js";
import { finiteNumber, isMain, projectDirectory, unique } from "./lib/runtime.js";

export const HISTORY_SLOTS = 5;
export const MIN_MEDIAN_YEARS = 3;

export const OUTPUT_FIELDS = [
  "ticker", "cik", "name", "sector", "subsector", "valuation_price", "price_date",
  "price_observations", "price_status", "price_reason", "filing_date", "period_end",
  "revenue_ttm", "operating_income_ttm", "net_income_ttm", "free_cash_flow_ttm",
  "cash_and_equivalents", "total_debt", "stockholders_equity", "shares_outstanding",
  "market_cap", "net_debt", "enterprise_value", "earnings_yield", "free_cash_flow_yield",
  "price_to_earnings", "price_to_fcf", "ev_to_operating_income", "operating_margin",
  "five_year_revenue_growth", "positive_fcf_years", "historical_fcf_years",
  "net_debt_to_fcf", "price_to_book", "five_year_median_fcf",
  "historical_net_income_years", "five_year_median_net_income", "calculation_status",
  "calculation_warnings", "snapshot_source_url", "price_source_url",
];

export const RAW_FINANCIAL_FIELDS = [
  "revenue_ttm", "operating_income_ttm", "net_income_ttm", "free_cash_flow_ttm",
  "cash_and_equivalents", "total_debt", "stockholders_equity", "shares_outstanding",
];

export function parseArgs(argv = process.argv.slice(2)) {
  const root = projectDirectory(import.meta.url);
  const { values } = parseNodeArgs({
    args: argv,
    options: {
      universe: { type: "string", default: `${root}/generated/universe.csv` },
      snapshot: { type: "string", default: `${root}/generated/data/latest_snapshot.csv` },
      prices: { type: "string", default: `${root}/generated/data/latest_prices.csv` },
      output: { type: "string", default: `${root}/generated/data/latest_screen.csv` },
      help: { type: "boolean", short: "h" },
    },
  });
  return {
    universe: resolve(values.universe),
    snapshot: resolve(values.snapshot),
    prices: resolve(values.prices),
    output: resolve(values.output),
    help: values.help ?? false,
  };
}

export function loadCsv(path, requiredFields) {
  const { fields, rows } = readCsv(path);
  const missing = [...requiredFields].filter((field) => !fields.includes(field)).sort();
  if (rows.length === 0 || missing.length) {
    throw new Error(`${path} is empty or missing required columns: ${missing.join(", ")}`);
  }
  return rows;
}

export function indexRows(rows, label) {
  const indexed = new Map();
  for (const row of rows) {
    const ticker = String(row.ticker ?? "").trim();
    if (!ticker) throw new Error(`${label} contains a row without a ticker.`);
    if (indexed.has(ticker)) throw new Error(`${label} contains duplicate ticker: ${ticker}`);
    indexed.set(ticker, row);
  }
  return indexed;
}

export function requireSameTickers(universe, indexed, label) {
  const universeTickers = new Set(universe.map((row) => row.ticker.trim()));
  const missing = [...universeTickers].filter((ticker) => !indexed.has(ticker)).sort();
  const extra = [...indexed.keys()].filter((ticker) => !universeTickers.has(ticker)).sort();
  if (missing.length || extra.length) {
    const details = [];
    if (missing.length) details.push(`missing ${missing.length} (${missing.slice(0, 5).join(", ")})`);
    if (extra.length) details.push(`extra ${extra.length} (${extra.slice(0, 5).join(", ")})`);
    throw new Error(`${label} ticker set does not match universe.csv: ${details.join("; ")}`);
  }
}

export function asNumber(value) {
  return finiteNumber(value);
}

function fixedWithoutZeros(value, decimals) {
  const text = value.toFixed(decimals).replace(/\.?0+$/, "");
  return text === "-0" || text === "" ? "0" : text;
}

export function formatAmount(value) {
  return value === null ? "" : fixedWithoutZeros(value, 6);
}

export function formatRatio(value) {
  if (value === null) return "";
  const rounded = Number(value.toPrecision(15));
  const absolute = Math.abs(rounded);
  const text = absolute > 0 && absolute < 1e-6
    ? rounded.toFixed(20).replace(/\.?0+$/, "")
    : rounded.toLocaleString("en-US", {
      useGrouping: false,
      maximumFractionDigits: 20,
    });
  return text === "-0" || text === "" ? "0" : text;
}

export function divide(numerator, denominator) {
  return numerator === null || denominator === null || denominator === 0
    ? null
    : numerator / denominator;
}

export function positiveDenominatorRatio(numerator, denominator) {
  return denominator === null || denominator <= 0 ? null : divide(numerator, denominator);
}

export function annualValues(snapshot, field) {
  const values = [];
  for (let slot = 0; slot < HISTORY_SLOTS; slot += 1) {
    const value = asNumber(snapshot[`${field}_fy${slot}`]);
    if (value !== null) values.push(value);
  }
  return values;
}

export function median(values) {
  const ordered = [...values].sort((left, right) => left - right);
  const middle = Math.floor(ordered.length / 2);
  return ordered.length % 2 ? ordered[middle] : (ordered[middle - 1] + ordered[middle]) / 2;
}

export function historicalMedian(values) {
  return values.length < MIN_MEDIAN_YEARS ? null : median(values);
}

export function revenueCagr(snapshot) {
  const newest = asNumber(snapshot.revenue_fy0);
  const oldest = asNumber(snapshot.revenue_fy4);
  const newestDate = parseDate(snapshot.annual_period_end_0);
  const oldestDate = parseDate(snapshot.annual_period_end_4);
  if (
    newest === null || oldest === null || newest <= 0 || oldest <= 0
    || !newestDate || !oldestDate || newestDate <= oldestDate
  ) return null;
  const years = daysBetween(oldestDate, newestDate) / 365.2425;
  if (years <= 0) return null;
  const growth = (newest / oldest) ** (1 / years) - 1;
  return Number.isFinite(growth) ? growth : null;
}

export function addInputWarning(warnings, field, value, requirePositive) {
  if (value === null) warnings.push(`missing_${field}`);
  else if (requirePositive && value <= 0) warnings.push(`nonpositive_${field}`);
}

export function buildRow(universe, snapshot, price) {
  const row = {
    ticker: universe.ticker.trim(),
    cik: String(universe.cik ?? "").trim(),
    name: String(universe.name ?? "").trim(),
    sector: String(universe.sector ?? "").trim(),
    subsector: String(universe.subsector ?? "").trim(),
    valuation_price: String(price.valuation_price ?? "").trim(),
    price_date: String(price.price_date ?? "").trim(),
    price_observations: String(price.observations ?? "").trim(),
    price_status: String(price.status ?? "").trim(),
    price_reason: String(price.reason ?? "").trim(),
    filing_date: String(snapshot.filing_date ?? "").trim(),
    period_end: String(snapshot.period_end ?? "").trim(),
    snapshot_source_url: String(snapshot.source_url ?? "").trim(),
    price_source_url: String(price.source_url ?? "").trim(),
  };
  for (const field of RAW_FINANCIAL_FIELDS) row[field] = String(snapshot[field] ?? "").trim();

  if (row.price_status !== "ok") {
    row.calculation_status = "excluded";
    row.calculation_warnings = row.price_reason || "price_unavailable";
    return Object.fromEntries(OUTPUT_FIELDS.map((field) => [field, row[field] ?? ""]));
  }

  const valuationPrice = asNumber(row.valuation_price);
  const shares = asNumber(row.shares_outstanding);
  const revenue = asNumber(row.revenue_ttm);
  const operatingIncome = asNumber(row.operating_income_ttm);
  const netIncome = asNumber(row.net_income_ttm);
  const freeCashFlow = asNumber(row.free_cash_flow_ttm);
  const cash = asNumber(row.cash_and_equivalents);
  const debt = asNumber(row.total_debt);
  const equity = asNumber(row.stockholders_equity);

  const warnings = [];
  for (const [field, value, requirePositive] of [
    ["valuation_price", valuationPrice, true], ["shares_outstanding", shares, true],
    ["revenue_ttm", revenue, true], ["operating_income_ttm", operatingIncome, true],
    ["net_income_ttm", netIncome, true], ["free_cash_flow_ttm", freeCashFlow, true],
    ["cash_and_equivalents", cash, false], ["total_debt", debt, false],
    ["stockholders_equity", equity, true],
  ]) addInputWarning(warnings, field, value, requirePositive);

  const marketCap = valuationPrice !== null && valuationPrice > 0 && shares !== null && shares > 0
    ? valuationPrice * shares : null;
  const netDebt = debt !== null && cash !== null ? debt - cash : null;
  const enterpriseValue = marketCap !== null && netDebt !== null ? marketCap + netDebt : null;
  const earningsYield = positiveDenominatorRatio(netIncome, marketCap);
  const freeCashFlowYield = positiveDenominatorRatio(freeCashFlow, marketCap);
  const priceToEarnings = positiveDenominatorRatio(marketCap, netIncome);
  const priceToFcf = positiveDenominatorRatio(marketCap, freeCashFlow);
  const evToOperatingIncome = enterpriseValue !== null && enterpriseValue > 0
    ? positiveDenominatorRatio(enterpriseValue, operatingIncome) : null;
  const operatingMargin = positiveDenominatorRatio(operatingIncome, revenue);
  const netDebtToFcf = positiveDenominatorRatio(netDebt, freeCashFlow);
  const priceToBook = positiveDenominatorRatio(marketCap, equity);

  const revenueGrowth = revenueCagr(snapshot);
  if (revenueGrowth === null) warnings.push("insufficient_revenue_history");
  const fcfHistory = annualValues(snapshot, "free_cash_flow");
  const netIncomeHistory = annualValues(snapshot, "net_income");
  const medianFcf = historicalMedian(fcfHistory);
  const medianNetIncome = historicalMedian(netIncomeHistory);
  if (medianFcf === null) warnings.push("insufficient_fcf_history");
  if (medianNetIncome === null) warnings.push("insufficient_net_income_history");

  Object.assign(row, {
    market_cap: formatAmount(marketCap),
    net_debt: formatAmount(netDebt),
    enterprise_value: formatAmount(enterpriseValue),
    earnings_yield: formatRatio(earningsYield),
    free_cash_flow_yield: formatRatio(freeCashFlowYield),
    price_to_earnings: formatRatio(priceToEarnings),
    price_to_fcf: formatRatio(priceToFcf),
    ev_to_operating_income: formatRatio(evToOperatingIncome),
    operating_margin: formatRatio(operatingMargin),
    five_year_revenue_growth: formatRatio(revenueGrowth),
    positive_fcf_years: fcfHistory.length ? fcfHistory.filter((value) => value > 0).length : "",
    historical_fcf_years: fcfHistory.length,
    net_debt_to_fcf: formatRatio(netDebtToFcf),
    price_to_book: formatRatio(priceToBook),
    five_year_median_fcf: formatAmount(medianFcf),
    historical_net_income_years: netIncomeHistory.length,
    five_year_median_net_income: formatAmount(medianNetIncome),
    calculation_status: warnings.length ? "partial" : "ok",
    calculation_warnings: unique(warnings).join(";"),
  });
  return Object.fromEntries(OUTPUT_FIELDS.map((field) => [field, row[field] ?? ""]));
}

export async function main(argv = process.argv.slice(2)) {
  try {
    const args = parseArgs(argv);
    if (args.help) {
      console.log("Usage: node scripts/build_screen.js [--universe PATH] [--snapshot PATH] [--prices PATH] [--output PATH]");
      return 0;
    }
    const universe = loadCsv(args.universe, new Set(["ticker", "cik", "name", "sector", "subsector"]));
    const snapshots = indexRows(
      loadCsv(args.snapshot, new Set(["ticker", ...RAW_FINANCIAL_FIELDS])),
      "latest_snapshot.csv",
    );
    const prices = indexRows(
      loadCsv(args.prices, new Set(["ticker", "valuation_price", "status", "reason"])),
      "latest_prices.csv",
    );
    requireSameTickers(universe, snapshots, "latest_snapshot.csv");
    requireSameTickers(universe, prices, "latest_prices.csv");
    const rows = universe.map((row) => buildRow(row, snapshots.get(row.ticker.trim()), prices.get(row.ticker.trim())));
    writeCsvAtomic(args.output, rows, OUTPUT_FIELDS);
    console.log(`Wrote ${rows.length} rows to ${args.output}`);
    for (const status of ["ok", "partial", "excluded"]) {
      const count = rows.filter((row) => row.calculation_status === status).length;
      if (count) console.log(`  ${status}: ${count}`);
    }
    return 0;
  } catch (error) {
    console.error(`Error: ${error.message}`);
    console.error("The existing screen file was not changed.");
    return 1;
  }
}

if (isMain(import.meta.url)) process.exitCode = await main();

File: scripts/build_screen.py
#!/usr/bin/env python3

"""Join universe, SEC fundamentals, and HF prices into latest_screen.csv.

The output keeps one row per security in universe.csv. Calculations are left
blank when required inputs are missing or when a valuation multiple would have
an economically meaningless non-positive denominator.

Usage:

    python3 scripts/build_screen.py
    python3 scripts/build_screen.py --output /tmp/latest_screen.csv

Only the standard Python library is used.
"""

from __future__ import annotations

import argparse
import csv
import math
import os
import statistics
import sys
import tempfile
from datetime import date
from decimal import Decimal, InvalidOperation, localcontext
from pathlib import Path
from typing import Any, Sequence


HISTORY_SLOTS = 5
MIN_MEDIAN_YEARS = 3

OUTPUT_FIELDS = [
    "ticker",
    "cik",
    "name",
    "sector",
    "subsector",
    "valuation_price",
    "price_date",
    "price_observations",
    "price_status",
    "price_reason",
    "filing_date",
    "period_end",
    "revenue_ttm",
    "operating_income_ttm",
    "net_income_ttm",
    "free_cash_flow_ttm",
    "cash_and_equivalents",
    "total_debt",
    "stockholders_equity",
    "shares_outstanding",
    "market_cap",
    "net_debt",
    "enterprise_value",
    "earnings_yield",
    "free_cash_flow_yield",
    "price_to_earnings",
    "price_to_fcf",
    "ev_to_operating_income",
    "operating_margin",
    "five_year_revenue_growth",
    "positive_fcf_years",
    "historical_fcf_years",
    "net_debt_to_fcf",
    "price_to_book",
    "five_year_median_fcf",
    "historical_net_income_years",
    "five_year_median_net_income",
    "calculation_status",
    "calculation_warnings",
    "snapshot_source_url",
    "price_source_url",
]

RAW_FINANCIAL_FIELDS = (
    "revenue_ttm",
    "operating_income_ttm",
    "net_income_ttm",
    "free_cash_flow_ttm",
    "cash_and_equivalents",
    "total_debt",
    "stockholders_equity",
    "shares_outstanding",
)


def parse_args(argv: Sequence[str] | None = None) -> argparse.Namespace:
    project_dir = Path(__file__).resolve().parent.parent
    parser = argparse.ArgumentParser(
        description="Build valuation metrics from the latest price and SEC snapshots."
    )
    parser.add_argument(
        "--universe",
        type=Path,
        default=project_dir / "generated" / "universe.csv",
        help="Universe CSV (default: ./generated/universe.csv).",
    )
    parser.add_argument(
        "--snapshot",
        type=Path,
        default=project_dir / "generated" / "data" / "latest_snapshot.csv",
        help="SEC snapshot CSV (default: ./generated/data/latest_snapshot.csv).",
    )
    parser.add_argument(
        "--prices",
        type=Path,
        default=project_dir / "generated" / "data" / "latest_prices.csv",
        help="Price snapshot CSV (default: ./generated/data/latest_prices.csv).",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=project_dir / "generated" / "data" / "latest_screen.csv",
        help="Output CSV (default: ./generated/data/latest_screen.csv).",
    )
    return parser.parse_args(argv)


def load_csv(path: Path, required_fields: set[str]) -> list[dict[str, str]]:
    if not path.exists():
        raise RuntimeError(f"Input file does not exist: {path}")
    with path.open(newline="", encoding="utf-8-sig") as handle:
        reader = csv.DictReader(handle)
        rows = list(reader)
        fields = set(reader.fieldnames or [])
    if not rows or not required_fields.issubset(fields):
        missing = ", ".join(sorted(required_fields - fields))
        raise RuntimeError(f"{path} is empty or missing required columns: {missing}")
    return rows


def index_rows(rows: list[dict[str, str]], label: str) -> dict[str, dict[str, str]]:
    indexed: dict[str, dict[str, str]] = {}
    for row in rows:
        ticker = (row.get("ticker") or "").strip()
        if not ticker:
            raise RuntimeError(f"{label} contains a row without a ticker.")
        if ticker in indexed:
            raise RuntimeError(f"{label} contains duplicate ticker: {ticker}")
        indexed[ticker] = row
    return indexed


def require_same_tickers(
    universe: list[dict[str, str]],
    indexed: dict[str, dict[str, str]],
    label: str,
) -> None:
    universe_tickers = {row["ticker"].strip() for row in universe}
    indexed_tickers = set(indexed)
    missing = sorted(universe_tickers - indexed_tickers)
    extra = sorted(indexed_tickers - universe_tickers)
    if missing or extra:
        details: list[str] = []
        if missing:
            details.append(f"missing {len(missing)} ({', '.join(missing[:5])})")
        if extra:
            details.append(f"extra {len(extra)} ({', '.join(extra[:5])})")
        raise RuntimeError(f"{label} ticker set does not match universe.csv: {'; '.join(details)}")


def as_decimal(value: Any) -> Decimal | None:
    text = str(value or "").strip()
    if not text:
        return None
    try:
        number = Decimal(text)
    except InvalidOperation:
        return None
    return number if number.is_finite() else None


def as_date(value: Any) -> date | None:
    text = str(value or "").strip()
    if not text:
        return None
    try:
        return date.fromisoformat(text[:10])
    except ValueError:
        return None


def format_amount(value: Decimal | None) -> str:
    if value is None:
        return ""
    text = f"{value:.6f}".rstrip("0").rstrip(".")
    return "0" if text in {"-0", ""} else text


def format_ratio(value: Decimal | None) -> str:
    if value is None:
        return ""
    # Preserve significant digits for both ordinary ratios and extreme audit
    # cases caused by malformed source share counts (for example, 0.00000003x).
    with localcontext() as context:
        context.prec = 15
        text = format(+value, "f")
    if "." in text:
        text = text.rstrip("0").rstrip(".")
    return "0" if text in {"-0", ""} else text


def divide(numerator: Decimal | None, denominator: Decimal | None) -> Decimal | None:
    if numerator is None or denominator is None or denominator == 0:
        return None
    with localcontext() as context:
        context.prec = 28
        return numerator / denominator


def positive_denominator_ratio(
    numerator: Decimal | None, denominator: Decimal | None
) -> Decimal | None:
    if denominator is None or denominator <= 0:
        return None
    return divide(numerator, denominator)


def annual_values(snapshot: dict[str, str], field: str) -> list[Decimal]:
    values: list[Decimal] = []
    for slot in range(HISTORY_SLOTS):
        value = as_decimal(snapshot.get(f"{field}_fy{slot}"))
        if value is not None:
            values.append(value)
    return values


def historical_median(values: list[Decimal]) -> Decimal | None:
    if len(values) < MIN_MEDIAN_YEARS:
        return None
    return statistics.median(values)


def revenue_cagr(snapshot: dict[str, str]) -> Decimal | None:
    newest = as_decimal(snapshot.get("revenue_fy0"))
    oldest = as_decimal(snapshot.get("revenue_fy4"))
    newest_date = as_date(snapshot.get("annual_period_end_0"))
    oldest_date = as_date(snapshot.get("annual_period_end_4"))
    if (
        newest is None
        or oldest is None
        or newest <= 0
        or oldest <= 0
        or newest_date is None
        or oldest_date is None
        or newest_date <= oldest_date
    ):
        return None
    years = (newest_date - oldest_date).days / 365.2425
    if years <= 0:
        return None
    growth = math.pow(float(newest / oldest), 1.0 / years) - 1.0
    if not math.isfinite(growth):
        return None
    return Decimal(str(growth))


def add_input_warning(
    warnings: list[str],
    field: str,
    value: Decimal | None,
    require_positive: bool,
) -> None:
    if value is None:
        warnings.append(f"missing_{field}")
    elif require_positive and value <= 0:
        warnings.append(f"nonpositive_{field}")


def build_row(
    universe: dict[str, str],
    snapshot: dict[str, str],
    price: dict[str, str],
) -> dict[str, Any]:
    row: dict[str, Any] = {
        "ticker": universe["ticker"].strip(),
        "cik": (universe.get("cik") or "").strip(),
        "name": (universe.get("name") or "").strip(),
        "sector": (universe.get("sector") or "").strip(),
        "subsector": (universe.get("subsector") or "").strip(),
        "valuation_price": (price.get("valuation_price") or "").strip(),
        "price_date": (price.get("price_date") or "").strip(),
        "price_observations": (price.get("observations") or "").strip(),
        "price_status": (price.get("status") or "").strip(),
        "price_reason": (price.get("reason") or "").strip(),
        "filing_date": (snapshot.get("filing_date") or "").strip(),
        "period_end": (snapshot.get("period_end") or "").strip(),
        "snapshot_source_url": (snapshot.get("source_url") or "").strip(),
        "price_source_url": (price.get("source_url") or "").strip(),
    }
    for field in RAW_FINANCIAL_FIELDS:
        row[field] = (snapshot.get(field) or "").strip()

    if row["price_status"] != "ok":
        row["calculation_status"] = "excluded"
        row["calculation_warnings"] = row["price_reason"] or "price_unavailable"
        return {field: row.get(field, "") for field in OUTPUT_FIELDS}

    valuation_price = as_decimal(row["valuation_price"])
    shares = as_decimal(row["shares_outstanding"])
    revenue = as_decimal(row["revenue_ttm"])
    operating_income = as_decimal(row["operating_income_ttm"])
    net_income = as_decimal(row["net_income_ttm"])
    free_cash_flow = as_decimal(row["free_cash_flow_ttm"])
    cash = as_decimal(row["cash_and_equivalents"])
    debt = as_decimal(row["total_debt"])
    equity = as_decimal(row["stockholders_equity"])

    warnings: list[str] = []
    add_input_warning(warnings, "valuation_price", valuation_price, True)
    add_input_warning(warnings, "shares_outstanding", shares, True)
    add_input_warning(warnings, "revenue_ttm", revenue, True)
    add_input_warning(warnings, "operating_income_ttm", operating_income, True)
    add_input_warning(warnings, "net_income_ttm", net_income, True)
    add_input_warning(warnings, "free_cash_flow_ttm", free_cash_flow, True)
    add_input_warning(warnings, "cash_and_equivalents", cash, False)
    add_input_warning(warnings, "total_debt", debt, False)
    add_input_warning(warnings, "stockholders_equity", equity, True)

    market_cap = (
        valuation_price * shares
        if valuation_price is not None
        and valuation_price > 0
        and shares is not None
        and shares > 0
        else None
    )
    net_debt = debt - cash if debt is not None and cash is not None else None
    enterprise_value = (
        market_cap + net_debt
        if market_cap is not None and net_debt is not None
        else None
    )

    earnings_yield = positive_denominator_ratio(net_income, market_cap)
    free_cash_flow_yield = positive_denominator_ratio(free_cash_flow, market_cap)
    price_to_earnings = positive_denominator_ratio(market_cap, net_income)
    price_to_fcf = positive_denominator_ratio(market_cap, free_cash_flow)
    ev_to_operating_income = (
        positive_denominator_ratio(enterprise_value, operating_income)
        if enterprise_value is not None and enterprise_value > 0
        else None
    )
    operating_margin = positive_denominator_ratio(operating_income, revenue)
    net_debt_to_fcf = positive_denominator_ratio(net_debt, free_cash_flow)
    price_to_book = positive_denominator_ratio(market_cap, equity)

    revenue_growth = revenue_cagr(snapshot)
    if revenue_growth is None:
        warnings.append("insufficient_revenue_history")

    fcf_history = annual_values(snapshot, "free_cash_flow")
    net_income_history = annual_values(snapshot, "net_income")
    median_fcf = historical_median(fcf_history)
    median_net_income = historical_median(net_income_history)
    if median_fcf is None:
        warnings.append("insufficient_fcf_history")
    if median_net_income is None:
        warnings.append("insufficient_net_income_history")

    row.update(
        {
            "market_cap": format_amount(market_cap),
            "net_debt": format_amount(net_debt),
            "enterprise_value": format_amount(enterprise_value),
            "earnings_yield": format_ratio(earnings_yield),
            "free_cash_flow_yield": format_ratio(free_cash_flow_yield),
            "price_to_earnings": format_ratio(price_to_earnings),
            "price_to_fcf": format_ratio(price_to_fcf),
            "ev_to_operating_income": format_ratio(ev_to_operating_income),
            "operating_margin": format_ratio(operating_margin),
            "five_year_revenue_growth": format_ratio(revenue_growth),
            "positive_fcf_years": (
                sum(value > 0 for value in fcf_history) if fcf_history else ""
            ),
            "historical_fcf_years": len(fcf_history),
            "net_debt_to_fcf": format_ratio(net_debt_to_fcf),
            "price_to_book": format_ratio(price_to_book),
            "five_year_median_fcf": format_amount(median_fcf),
            "historical_net_income_years": len(net_income_history),
            "five_year_median_net_income": format_amount(median_net_income),
            "calculation_status": "ok" if not warnings else "partial",
            "calculation_warnings": ";".join(dict.fromkeys(warnings)),
        }
    )
    return {field: row.get(field, "") for field in OUTPUT_FIELDS}


def write_atomically(output_path: Path, rows: list[dict[str, Any]]) -> None:
    output_path.parent.mkdir(parents=True, exist_ok=True)
    temporary_path: Path | None = None
    try:
        with tempfile.NamedTemporaryFile(
            mode="w",
            encoding="utf-8",
            newline="",
            prefix=f".{output_path.name}.",
            suffix=".tmp",
            dir=output_path.parent,
            delete=False,
        ) as temporary_file:
            temporary_path = Path(temporary_file.name)
            writer = csv.DictWriter(
                temporary_file,
                fieldnames=OUTPUT_FIELDS,
                extrasaction="ignore",
                lineterminator="\n",
            )
            writer.writeheader()
            writer.writerows(rows)
            temporary_file.flush()
            os.fsync(temporary_file.fileno())

        os.replace(temporary_path, output_path)
        temporary_path = None
    finally:
        if temporary_path is not None:
            temporary_path.unlink(missing_ok=True)


def main(argv: Sequence[str] | None = None) -> int:
    args = parse_args(argv)
    try:
        universe = load_csv(
            args.universe.resolve(),
            {"ticker", "cik", "name", "sector", "subsector"},
        )
        snapshots = index_rows(
            load_csv(
                args.snapshot.resolve(),
                {"ticker", *RAW_FINANCIAL_FIELDS},
            ),
            "latest_snapshot.csv",
        )
        prices = index_rows(
            load_csv(
                args.prices.resolve(),
                {"ticker", "valuation_price", "status", "reason"},
            ),
            "latest_prices.csv",
        )
        require_same_tickers(universe, snapshots, "latest_snapshot.csv")
        require_same_tickers(universe, prices, "latest_prices.csv")

        rows = [
            build_row(row, snapshots[row["ticker"].strip()], prices[row["ticker"].strip()])
            for row in universe
        ]
        write_atomically(args.output.resolve(), rows)
    except (OSError, RuntimeError) as error:
        print(f"Error: {error}", file=sys.stderr)
        print("The existing screen file was not changed.", file=sys.stderr)
        return 1

    statuses: dict[str, int] = {}
    for row in rows:
        status = str(row["calculation_status"])
        statuses[status] = statuses.get(status, 0) + 1
    print(f"Wrote {len(rows)} rows to {args.output.resolve()}")
    for status in ("ok", "partial", "excluded"):
        if statuses.get(status):
            print(f"  {status}: {statuses[status]}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
